lib/gen_data/sorted_list.ex

```diff
@@ -0,0 +1,107 @@
+defmodule GenData.SortedList do
+  @moduledoc """
+  A sequential data structure that remains sorted.
+
+  Default sorting uses the `GenData.Comparable` instances of its members, or custom comparison can
+  be specified.
+
+  """
+
+  alias GenData.{Macros, Comparable}
+  require Macros
+
+  Macros.require_impls [Comparable]
+
+
+
+  ################################################################################
+  # Types
+  ################################################################################
+
+  @type comparator :: ((any, any) -> boolean)
+
+  @opaque t(contents) :: %__MODULE__{list: [contents], comparator: comparator}
+  @type t :: t(any)
+
+  defstruct [
+    list: [],
+    comparator: nil,
+  ]
+
+
+
+  ################################################################################
+  # Public API
+  ################################################################################
+
+  @doc """
+  Creates a sorted list from an enumerable.
+
+  ## Examples
+
+      iex> GenData.SortedList.new([1, 3, 2])
+      #GenData.SortedList<[1, 2, 3]>
+
+      iex> GenData.SortedList.new([1, 2, 3, 2])
+      #GenData.SortedList<[1, 2, 2, 3]>
+
+  """
+  @spec new(Enumerable.t, options) :: t
+    when options: [{:comparator, comparator}]
+  def new(enumerable \\ [], options \\ []) do
+    comparator =
+      Keyword.get(options, :comparator, &Comparable.compare(&1, &2) == :lt)
+
+    %__MODULE__{
+      list: Enum.sort(enumerable, comparator),
+      comparator: comparator,
+    }
+  end
+
+
+  @doc """
+  Add a new element to the given sorted list.
+
+  ## Examples
+
+    iex> list = GenData.SortedList.new([3, 4, 1])
+    iex> GenData.SortedList.put(list, 2)
+    #GenData.SortedList<[1, 2, 3, 4]>
+
+  """
+  @spec put(t, any) :: t
+  def put(%__MODULE__{list: contents, comparator: comparator} = list, x) do
```

Comment:
`put` feels like a map / set backed operation. Since this module describes itself as a `List` I am wondering if we should perhaps call it `insert` or `push`. Alternatively, we could call it a `SortedCollection` to avoid implying data structure; that being said, I prefer knowing the data structure I am using for the sake of intuiting performance characteristics about what I am using (ie. perhaps we should have SortedList and SortedTree, etc).